### ISSUE TITLE
cf create-buildpack fails with 500 error - cloud_controller_ng.log shows Permission denied @ rb_sysopen using CF 253 and local NFS blobstore

### ISSUE THREAD

**Issue by chrisahl** [cloudfoundry/cloud_controller_ng#934]
## Issue

Using CF 253, `cf create-buildpack` is returning a 500 error.  The cloud_controller_ng.log shows error:  Permission denied @ rb_sysopen

## Context

cloud_controller_ng.log shows:
```{"timestamp":1507811621.6225111,"message":"cc.dispatch","log_level":"debug","source":"cc.api","data":{"request_guid":"df04b32a-3512-413c-b26a-a5b19a7a89af","endpoint":"list","args":[]},"thread_id":47206024060480,"fiber_id":47206033634340,"process_id":27277,"file":"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/app/controllers/base/base_controller.rb","lineno":78,"method":"dispatch"}
{"timestamp":1507811621.162408,"message":"cp_to_blobstore-retry","log_level":"debug","source":"cc.blobstore","data":{"request_guid":"b2bce9be-4594-4bfd-4977-f2d2b4ea642d::d83a2926-2784-49a0-b16d-4e0be19d539c","error":"Input/output error @ fptr_finalize - /var/vcap/nfs/shared/cc-buildpacks/8d/3c/8d3ccb97-5fb2-4885-acb6-0c31a968c0ce_21be9f1ab8ce3cc605c07bc6196aa970a3cf83255b3f4ce98fe5d4673e4a9edf","remaining_retries":2,"args":{"source_path":"/var/vcap/data/cloud_controller_ng/tmp/uploads/0000000001","destination_key":"8d3ccb97-5fb2-4885-acb6-0c31a968c0ce_21be9f1ab8ce3cc605c07bc6196aa970a3cf83255b3f4ce98fe5d4673e4a9edf"}},"thread_id":47206024060760,"fiber_id":47206033866320,"process_id":27277,"file":"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/retryable_client.rb","lineno":131,"method":"rescue in with_retries"}
{"timestamp":1507811621.6277585,"message":"blobstore.cp-start","log_level":"info","source":"cc.blobstore","data":{"request_guid":"b2bce9be-4594-4bfd-4977-f2d2b4ea642d::d83a2926-2784-49a0-b16d-4e0be19d539c","destination_key":"8d3ccb97-5fb2-4885-acb6-0c31a968c0ce_21be9f1ab8ce3cc605c07bc6196aa970a3cf83255b3f4ce98fe5d4673e4a9edf","source_path":"/var/vcap/data/cloud_controller_ng/tmp/uploads/0000000001","bucket":"cc-buildpacks"},"thread_id":47206024060760,"fiber_id":47206033866320,"process_id":27277,"file":"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/fog/fog_client.rb","lineno":53,"method":"cp_to_blobstore"}
{"timestamp":1507811621.6300044,"message":"cp_to_blobstore-retry","log_level":"debug","source":"cc.blobstore","data":{"request_guid":"b2bce9be-4594-4bfd-4977-f2d2b4ea642d::d83a2926-2784-49a0-b16d-4e0be19d539c","error":"Permission denied @ rb_sysopen - /var/vcap/nfs/shared/cc-buildpacks/8d/3c/8d3ccb97-5fb2-4885-acb6-0c31a968c0ce_21be9f1ab8ce3cc605c07bc6196aa970a3cf83255b3f4ce98fe5d4673e4a9edf","remaining_retries":1,"args":{"source_path":"/var/vcap/data/cloud_controller_ng/tmp/uploads/0000000001","destination_key":"8d3ccb97-5fb2-4885-acb6-0c31a968c0ce_21be9f1ab8ce3cc605c07bc6196aa970a3cf83255b3f4ce98fe5d4673e4a9edf"}},"thread_id":47206024060760,"fiber_id":47206033866320,"process_id":27277,"file":"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/retryable_client.rb","lineno":131,"method":"rescue in with_retries"}
{"timestamp":1507811621.630315,"message":"blobstore.cp-start","log_level":"info","source":"cc.blobstore","data":{"request_guid":"b2bce9be-4594-4bfd-4977-f2d2b4ea642d::d83a2926-2784-49a0-b16d-4e0be19d539c","destination_key":"8d3ccb97-5fb2-4885-acb6-0c31a968c0ce_21be9f1ab8ce3cc605c07bc6196aa970a3cf83255b3f4ce98fe5d4673e4a9edf","source_path":"/var/vcap/data/cloud_controller_ng/tmp/uploads/0000000001","bucket":"cc-buildpacks"},"thread_id":47206024060760,"fiber_id":47206033866320,"process_id":27277,"file":"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/fog/fog_client.rb","lineno":53,"method":"cp_to_blobstore"}
{"timestamp":1507811621.6316981,"message":"cp_to_blobstore-retry","log_level":"debug","source":"cc.blobstore","data":{"request_guid":"b2bce9be-4594-4bfd-4977-f2d2b4ea642d::d83a2926-2784-49a0-b16d-4e0be19d539c","error":"Permission denied @ rb_sysopen - /var/vcap/nfs/shared/cc-buildpacks/8d/3c/8d3ccb97-5fb2-4885-acb6-0c31a968c0ce_21be9f1ab8ce3cc605c07bc6196aa970a3cf83255b3f4ce98fe5d4673e4a9edf","remaining_retries":0,"args":{"source_path":"/var/vcap/data/cloud_controller_ng/tmp/uploads/0000000001","destination_key":"8d3ccb97-5fb2-4885-acb6-0c31a968c0ce_21be9f1ab8ce3cc605c07bc6196aa970a3cf83255b3f4ce98fe5d4673e4a9edf"}},"thread_id":47206024060760,"fiber_id":47206033866320,"process_id":27277,"file":"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/retryable_client.rb","lineno":131,"method":"rescue in with_retries"}
{"timestamp":1507811621.6325166,"message":"Request failed: 500: {\"error_code\"=>\"UnknownError\", \"description\"=>\"An unknown error occurred.\", \"code\"=>10001, \"test_mode_info\"=>{\"description\"=>\"Permission denied @ rb_sysopen - /var/vcap/nfs/shared/cc-buildpacks/8d/3c/8d3ccb97-5fb2-4885-acb6-0c31a968c0ce_21be9f1ab8ce3cc605c07bc6196aa970a3cf83255b3f4ce98fe5d4673e4a9edf\", \"error_code\"=>\"CF-EACCES\", \"backtrace\"=>[\"/var/vcap/packages/ruby-2.3/lib/ruby/2.3.0/fileutils.rb:1393:in `initialize'\", \"/var/vcap/packages/ruby-2.3/lib/ruby/2.3.0/fileutils.rb:1393:in `open'\", \"/var/vcap/packages/ruby-2.3/lib/ruby/2.3.0/fileutils.rb:1393:in `block in copy_file'\", \"/var/vcap/packages/ruby-2.3/lib/ruby/2.3.0/fileutils.rb:1392:in `open'\", \"/var/vcap/packages/ruby-2.3/lib/ruby/2.3.0/fileutils.rb:1392:in `copy_file'\", \"/var/vcap/packages/ruby-2.3/lib/ruby/2.3.0/fileutils.rb:486:in `copy_file'\", \"/var/vcap/packages/ruby-2.3/lib/ruby/2.3.0/fileutils.rb:403:in `block in cp'\", \"/var/vcap/packages/ruby-2.3/lib/ruby/2.3.0/fileutils.rb:1571:in `block in fu_each_src_dest'\", \"/var/vcap/packages/ruby-2.3/lib/ruby/2.3.0/fileutils.rb:1587:in `fu_each_src_dest0'\", \"/var/vcap/packages/ruby-2.3/lib/ruby/2.3.0/fileutils.rb:1569:in `fu_each_src_dest'\", \"/var/vcap/packages/ruby-2.3/lib/ruby/2.3.0/fileutils.rb:402:in `cp'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/fog-local-0.3.1/lib/fog/storage/local/models/file.rb:98:in `save'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/fog-core-1.43.0/lib/fog/core/collection.rb:51:in `create'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/fog/fog_client.rb:70:in `block in cp_to_blobstore'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/fog/fog_client.rb:57:in `open'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/fog/fog_client.rb:57:in `cp_to_blobstore'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/retryable_client.rb:46:in `block in cp_to_blobstore'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/retryable_client.rb:126:in `with_retries'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/retryable_client.rb:40:in `cp_to_blobstore'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/fog/error_handling_client.rb:49:in `block in cp_to_blobstore'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/fog/error_handling_client.rb:55:in `error_handling'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/blobstore/fog/error_handling_client.rb:49:in `cp_to_blobstore'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/upload_buildpack.rb:22:in `upload_buildpack'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/app/controllers/runtime/buildpack_bits_controller.rb:33:in `upload'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/app/controllers/base/base_controller.rb:80:in `dispatch'\", \"/var/vcap/data/packages/cloud_controller_ng/50f1e9080429f03d8fbdaede647d08e8a1bac2f5/cloud_controller_ng/lib/cloud_controller/rest_controller/routes.rb:16:in `block in define_route'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1610:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1610:in `block in compile!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:975:in `block (3 levels) in route!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:994:in `route_eval'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:975:in `block (2 levels) in route!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1015:in `block in process_route'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1013:in `catch'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1013:in `process_route'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:973:in `block in route!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:972:in `each'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:972:in `route!'\", ........

```

I can see that the buildpack gets uploaded to /var/vcap/data/cloud_controller_ng/tmp/uploads and then gets moved to the local nfs /var/vcap/nfs/shared/cc-buildpacks/8d/3c

```
# ls -al /var/vcap/nfs/shared/cc-buildpacks/8d/3c/
total 481864
drwxr-xr-x 2 vcap vcap      4096 Oct 12 12:33 .
drwxr-xr-x 3 vcap vcap      4096 Oct 12 12:33 ..
-r-------- 1 vcap vcap 493419433 Oct 12 12:33 8d3ccb97-5fb2-4885-acb6-0c31a968c0ce_21be9f1ab8ce3cc605c07bc6196aa970a3cf83255b3f4ce98fe5d4673e4a9edf
``` 

However there is some error during this process that causes an error.

## Steps to Reproduce

- Install CF 253, use local NFS blobstore
- cf create-buildpack <buildpack name> <buildpack zip> <position>
- Error displayed 

```
FAILED
Server error, status code: 500, error code: 10001, message: An unknown error occurred.
```

## Expected result

The buildpack should successfully be loaded and saved.

## Current result

buildpack cannot be created via cf command line

## Possible Fix

Similar item reported in #897, but that was on a newer CF version

**Comment by gcoon151** [comment]
Keeping issue updated with slack #capi discussion 

> Unfortunately I couldn't reproduce this with `cf create-buildpack` nor with `cf update-buildpack`. Could you provide a step-by-step repro?  I also  tried this both with and without our fix for issue 897 --@ericp

**Comment by chrisahl** [comment]
Does any of this depend on the umask value for root or vcap?  What should the umask value be on the vm cloud_controller_ng and debian_nfs_server runs on?

**Comment by chrisahl** [comment]
We tried updating a file with the changes for https://github.com/cloudfoundry/cloud_controller_ng/commit/ef4ec09a20f4984ca4fed51c3927474dd1d6a01e but it did not help

**Comment by chrisahl** [comment]
One thing we tried was changing so that the debian_nfs_server was not used by buildpacks.  We used a local directory.  Basically `cc:`, used /var/vcap/data.
```
    buildpacks:
      blobstore_type: fog
      buildpack_directory_key: cc-buildpacks
      cdn: null
      fog_connection:
        #local_root: /var/vcap/nfs/shared
        local_root: /var/vcap/data
        provider: Local
```

This worked for creating buildpacks.   Any ideas on what we should do to debug what seems to be a debian_nfs_server issue?

**Comment by ljfranklin** [comment]
@chrisahl @gcoon151 we've got a story coming up to get a CI env setup with NFS: https://www.pivotaltracker.com/story/show/150375975. Once that's done we can dig a little deeper into this issue.

I should also note that while CC does support NFS as a blobstore via the [nfs_mounter](https://github.com/cloudfoundry/capi-release/tree/develop/jobs/nfs_mounter), I believe we consider the [debian_nfs_server](https://github.com/cloudfoundry/capi-release/tree/develop/jobs/debian_nfs_server) to be somewhat deprecated. We don't run tests against it and can't vouch for whether it still works. I'm not sure whether there is another bosh release of NFS server that is better supported.

**Comment by chrisahl** [comment]
@ljfranklin Based on recent testing, we have narrowed the error down to how the vm running cloud_controller_ng mounts NFS.  The mount is coming from:

/var/vcap/data/jobs/nfs_mounter/6f7ef273c98b6d88b25c9c886422675591f049a5/bin/handle_nfs_blobstore.sh:  mount --verbose -o timeo=10,intr,lookupcache=positive,soft -t nfs 0.debian-nfs-server.default.bluemix.microbosh:/var/vcap/store $NFS_SHARE
      
We are experimenting with trying:
- increasing the timeo value
- changing from soft to hard
Both have given better results.  We would be interested in what your team's preference would be for a code change.    By this afternoon we hope to have more definite results after additional testing.

**Comment by chrisahl** [comment]
We have finished our testing of a patch for `cf create-buildpacks`.   We want the file handle_nfs_blobstore.sh (e.g. /var/vcap/data/jobs/nfs_mounter/6f7ef273c98b6d88b25c9c886422675591f049a5/bin/handle_nfs_blobstore.sh) to be changed so that the NFS mount uses `timeo=100` instead of `timeo=10`.   

This change produced 100% success for us.

**Comment by ljfranklin** [comment]
@chrisahl interesting, looks like `timeo` param for NFS mount is in _deciseconds_ instead of seconds: https://linux.die.net/man/5/nfs. The default value in the docs is 600 (60 seconds) before retrying a request. I can definitely imagine a value of `10` being too low even in a fairly healthy network.

As to changing from `hard` to `soft`, we had explicitly changed that mode as part of this story: https://www.pivotaltracker.com/story/show/121711865. As we checksum all the blobs before putting them in the blobstore, the possible data integrity issues mentioned in the docs around `soft` mode should be mitigated. When `hard` mode was enabled users could get their deployments wedged in strange states if the NFS server was ever unavailable.

**Comment by cwlbraa** [comment]
fixed here: 254fbca
